port/cpl_vsisimple.cpp

```diff
@@ -1371,20 +1372,107 @@ GIntBig CPLGetPhysicalRAM(void)
-    // If no limitation, on 64 bit, 9223372036854771712 is returned.
-    FILE *f = fopen("/sys/fs/cgroup/memory/memory.limit_in_bytes", "rb");
-    if (f)
+#ifdef __linux
+    char szGroupName[256];
+    bool bFromMemory = false;
+    szGroupName[0] = 0;
     {
-        char szBuffer[32];
-        const int nRead =
-            static_cast<int>(fread(szBuffer, 1, sizeof(szBuffer) - 1, f));
-        szBuffer[nRead] = 0;
+        FILE *f = fopen("/proc/self/cgroup", "rb");
+        char szLine[256];
+        // Find line like "6:memory:/user.slice/user-1000.slice/user@1000.service"
+        // and store "/user.slice/user-1000.slice/user@1000.service" in
+        // szMemoryPath for cgroup V1 or single line "0::/...." for cgroup V2.
+        while (fgets(szLine, sizeof(szLine), f))
+        {
+            const char *pszMemory = strstr(szLine, ":memory:");
+            if (pszMemory)
+            {
+                bFromMemory = true;
+                strcpy(szGroupName, pszMemory + strlen(":memory:"));
+                char *pszEOL = strchr(szGroupName, '\n');
+                if (pszEOL)
+                    *pszEOL = '\0';
+                break;
+            }
+            if (strncmp(szLine, "0::", strlen("0::")) == 0)
+            {
+                strcpy(szGroupName, szLine + strlen("0::"));
+                char *pszEOL = strchr(szGroupName, '\n');
+                if (pszEOL)
+                    *pszEOL = '\0';
+                break;
+            }
+        }
         fclose(f);
-        const GUIntBig nLimit = CPLScanUIntBig(szBuffer, nRead);
-        nVal =
-            static_cast<GIntBig>(std::min(static_cast<GUIntBig>(nVal), nLimit));
     }
+    if (szGroupName[0])
+    {
+        char szFilename[256 + 64];
+        if (bFromMemory)
+        {
+            // cgroup V1
+            // Read memory.limit_in_byte in the whole szGroupName hierarchy
+            while (true)
+            {
+                snprintf(szFilename, sizeof(szFilename),
+                         "/sys/fs/cgroup/memory/%s/memory.limit_in_bytes",
+                         szGroupName);
+                FILE *f = fopen(szFilename, "rb");
+                if (f)
+                {
+                    // If no limitation, on 64 bit, 9223372036854771712 is returned.
+                    char szBuffer[32];
+                    const int nRead = static_cast<int>(
+                        fread(szBuffer, 1, sizeof(szBuffer) - 1, f));
+                    szBuffer[nRead] = 0;
+                    fclose(f);
+                    const GUIntBig nLimit = CPLScanUIntBig(szBuffer, nRead);
+                    nVal = static_cast<GIntBig>(
+                        std::min(static_cast<GUIntBig>(nVal), nLimit));
+                }
+                char *pszLastSlash = strrchr(szGroupName, '/');
+                if (!pszLastSlash)
+                    break;
+                *pszLastSlash = '\0';
+            }
+        }
+        else
```

Comment:
Nit: might be worth trying V2 first, since it's going to be the most used path in the future (if not already).

Reply:
the logic (possibly flawed) is that if there is a ":memory:" line in "/proc/self/cgroup" then the V1 code path is taken. And if there's only a "0::" line, then V2 code path is taken.

Reply:
That's right, sorry.